okhttp/src/main/java/io/grpc/okhttp/OkHttpProtocolNegotiator.java

```diff
@@ -152,31 +179,61 @@ public String negotiate(SSLSocket sslSocket, String hostname, List<Protocol> pro
     /**
      * Override {@link Platform}'s configureTlsExtensions for Android older than 5.0, since OkHttp
      * (2.3+) only support such function for Android 5.0+.
+     *
+     * <p>Note: Prior to Android Q, the standard way of accessing some Conscrypt features was to
+     * use reflection to call hidden APIs. Beginning in Q, there is public API for all of these
+     * features. We attempt to use the public API where possible, but also still call the
+     * hidden versions to continue to support old versions of Conscrypt that might be bundled with
+     * apps or third-party TLS providers that might have taken advantage of being able to
+     * duck-type their way into compatibility.
      */
     @Override
     protected void configureTlsExtensions(
         SSLSocket sslSocket, String hostname, List<Protocol> protocols) {
       // Enable SNI and session tickets.
       if (hostname != null) {
-        SET_USE_SESSION_TICKETS.invokeOptionalWithoutCheckedException(sslSocket, true);
+        try {
+          if (sslSocketsIsSupportedSocket != null
+              && (boolean) sslSocketsIsSupportedSocket.invoke(null, sslSocket)) {
+            sslSocketsSetUseSessionTickets.invoke(null, sslSocket, true);
+          } else {
+            SET_USE_SESSION_TICKETS.invokeOptionalWithoutCheckedException(sslSocket, true);
+          }
+        } catch (IllegalAccessException ignored) {
+        } catch (InvocationTargetException ignored) {
+        }
         SET_HOSTNAME.invokeOptionalWithoutCheckedException(sslSocket, hostname);
       }
 
-      Object[] parameters = {Platform.concatLengthPrefixed(protocols)};
-      if (platform.getTlsExtensionType() == TlsExtensionType.ALPN_AND_NPN) {
-        SET_ALPN_PROTOCOLS.invokeWithoutCheckedException(sslSocket, parameters);
-      }
-
-      if (platform.getTlsExtensionType() != TlsExtensionType.NONE) {
-        SET_NPN_PROTOCOLS.invokeWithoutCheckedException(sslSocket, parameters);
-      } else {
-        throw new RuntimeException("We can not do TLS handshake on this Android version, please"
-            + " install the Google Play Services Dynamic Security Provider to use TLS");
+      // Enable ALPN and NPN if necessary.
+      SSLParameters sslParams = sslSocket.getSSLParameters();
+      SET_APPLICATION_PROTOCOLS
+          .invokeOptionalWithoutCheckedException(sslParams, (Object) protocolIds(protocols));
+      sslSocket.setSSLParameters(sslParams);
+      if (!isPlatformSocket(sslSocket) || !SET_APPLICATION_PROTOCOLS.isSupported(sslParams)) {
```

Comment:
I'm concerned about the `isPlatformSocket()` check. That seems brittle, especially since getting this wrong will spew ugly errors. If anything, it seems we should check for the gmscore version and apply special behavior in only that case.

In general, we want to use the new API, and only the new API, everywhere it is possible. We want the old API to naturally pass out of existence and know we aren't calling it at all on newer code.

Reply:
> we want to use the new API, and only the new API, everywhere it is possible.

That was my first thought as well. It was something like
```java
if (SET_APPLICATION_PROTOCOLS.isSupported(sslParams)) {
  // invoke SET_APPLICATION_PROTOCOLS on sslParams and sslSocket.setSSLParameters(sslParams)
} else {
  // invoke SET_ALPN_PROTOCOLS/SET_NPN_PROTOCOLS
}

```
However, when using bundled Conscrypt (i.e., `org.conscrypt`) on Android Q+, this doesn't work. We would still need to invoke SET_ALPN_PROTOCOLS/SET_NPN_PROTOCOLS.